fix(upgrade): $attrs and $tranclude dependencies for upgrade component

FDIM opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
When upgrading angularjs component, $attrs and $transclude dependencies are not provided as they are passed as locals.
Workaround is not really possible either as you can't get a reference to these locals from $injector and do something different for upgraded component without changing previous behavior.

Issue Number: #22824


## What is the new behavior?
$transclude is polyfilled, while $attrs is still left unsupported but provided as undefined

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

gkalpak commented on packages/upgrade/src/common/upgrade_helper.ts:
```diff
@@ -100,10 +100,13 @@ export class UpgradeHelper {
     // Quoted properties below so that this code can be optimized with Closure Compiler.
-    const locals = {'$scope': $scope, '$element': this.$element};
+    const locals =
+        {'$scope': $scope, '$element': this.$element, $attrs: $attrs, $transclude: $transclude};
```
If we support this in controllers, we should also support `$transclude` in [pre-/post-linking functions][1].
Also, we should pass the same value for `$attrs` in both.

[1]: https://github.com/angular/angular/blob/0bede54b2de0f3014a16fd59803858a3ac38b2f6/packages/upgrade/src/static/upgrade_component.ts#L173-L183

gkalpak commented on packages/upgrade/src/common/upgrade_helper.ts:
```diff
@@ -201,6 +204,24 @@ export class UpgradeHelper {
+    // Expose `isSlotFilled` helper function
+    transclude.isSlotFilled = (slotName) => !!slots[slotName];
+    return transclude;
+  }
```
This implementation is slightly different that the one in AngularJS (see [here][1]).
If we add this feature, we need to make it work as close as possible to plain AngularJS.

[1]: https://github.com/angular/angular.js/blob/c0adcc3a4b561260d0f1ff6b834ca940bf1971a9/src/ng/compile.js#L2859-L2892

FDIM replied:
true, I'll update it if we agree that the feature should be supported - The only thing I realized after doing is that it would simply not work for my case at work - So I'd be happy even with `undefined` :D

gkalpak reviewed (commented):
I am not strongly opposed, but a little bit sceptical (see https://github.com/angular/angular/issues/22824#issuecomment-374262127).

@petebacondarwin, wdyt?

FDIM committed:
fix(upgrade): $attrs and $tranclude dependencies for upgrade component

$transclude is polyfilled, while $attrs is still not supported and provided as undefined

Closes #22824

petebacondarwin commented:
Sounds like we should just aim for `undefined` in both cases, to unblock @FDIM - then later we could consider more support.

gkalpak commented:
Should we pass `'NOT_SUPPORTED'` as we do for linking functions? It has pros and cons 😕

<sub>BTW, I do believe we should be able to reasonably support `$transclude`, considering that we only need to support a subset of the features (e.g. these components can't have `transclude: 'element'`, etc.).</sub>

gkalpak commented:
Hm...thinking about it again, I am not sure it is a good idea to "pretend" to support something, only to avoid breaking people that "pretend" to depend on it :grin:

I mean, since we don't really support `$attrs` and `$transclude`(?), it is better to error when people try to inject them, rather than pass a dummy value and fail when they try to use it; just so that people that inject the values but don't use them are fine.

It is much easier for the latter group of people to either modify their implementations or create a alternative implementation that does not rely on the features. Or even provide the dummy/undefined fallback values through the injector.

I prefer failing faster rather than later :smiley:

##
I don't feel too strongly though, so if you two think it is a good idea, then it works for me :grin:

FDIM commented:
We have refactored/upgraded our components and no longer depend on this and seeing that there were no interests/complains from others for this feature / edge case, I'd say @gkalpak that you can close this PR if you agree.
